### ISSUE TITLE
IL2CPP error zenject 6.1.1 unity 2018.2

### ISSUE THREAD

**Issue by nievesj** [ssannandeji/Zenject-2019#503]
I'm getting this error on 2018.2 when building for ios, works perfectly fine on 2018.1. Nothing else was changed other than updating unity.

 .Net 4.6 on both unity versions.

IL2CPP error for field 'Zenject.DiContainer Zenject.KeyedFactoryBase`2::_container' in KeyedFactory.cs:11
Additional information: attempting to call method 'System.Void Zenject.InjectAttribute::.ctor()' that does not exist

**Comment by josencv** [comment]
Similar error here:

Unity version: 2018.2
Zenject version: 6.1.1
.NET version: 4.x

```
IL2CPP error for method 'System.Void Zenject.SignalBus::.ctor(System.Collections.Generic.List`1<Zenject.SignalDeclaration>,Zenject.SignalBus,Zenject.ZenjectSettings,Zenject.SignalSubscription/Pool)' in SignalBus.cs:18
Additional information: attempting to call method 'System.Void Zenject.InjectAttribute::.ctor()' that does not exist
```

**Comment by nievesj** [comment]
Reverted to Zenject 5.5.1 and that works with Unity 2018.2 and IL2CPP builds, I'll try updating zenject again when there's an update.

**Comment by quabug** [comment]
same here:
Unity version: 2018.2
Zenject version: 6.1.1
.Net version: standard 2.0

```IL2CPP error for method 'System.Void Zenject.DisposableManager::.ctor(System.Collections.Generic.List`1<System.IDisposable>,System.Collections.Generic.List`1<ModestTree.Util.ValuePair`2<System.Type,System.Int32>>,System.Collections.Generic.List`1<Zenject.ILateDisposable>,System.Collections.Generic.List`1<ModestTree.Util.ValuePair`2<System.Type,System.Int32>>)' in DisposableManager.cs:10
Additional information: attempting to call method 'System.Void Zenject.InjectAttribute::.ctor()' that does not exist```

**Comment by chung-ltkgame** [comment]
I have same issue with Unity 2018.2 and build with il2cpp
I checked in my project, and issue resolve when remove attribute Inject with Id
My issue same @quabug

**Comment by quabug** [comment]
@chung-ltkgame I just found a simple workaround by add a `link.xml` into your `Assets`(or any subdirectory of `Assets`).
```
<linker>
       <assembly fullname="Zenject-usage" preserve="all"/>
</linker>
```
see https://docs.unity3d.com/Manual/IL2CPP-BytecodeStripping.html for more info

**Comment by chung-ltkgame** [comment]
Update: I watched commits on master branh and now, I resolved issue with commit id f69a9b856cfe138718a5ae1bb71404f9b5c2385b.
Edit abstract class InjectAttributeBase extend System.Attribute instead of extend PreserveAttribute

**Comment by svermeulen** [comment]
It's a unity bug which I reported here and they confirmed:  https://fogbugz.unity3d.com/default.asp?1060608_jbo50fdc7r0p9gbi

For now I'll update the 2018.2 release to have the change f69a9b8 but that change could also create issues so not ideal.  Or maybe try adding a link file like @quabug suggests instead

**Comment by chung-ltkgame** [comment]
I have new issue, my library have dependency with Zenject. If I build library to dll and push to UnityProject then Inject is not working, but if I push source code library to Unity Project then Inject is working perfect

**Comment by svermeulen** [comment]
@chung-ltkgame Can you create a new github issue with more detail?  That sounds unrelated to this one

**Comment by svermeulen** [comment]
This issue is fixed now in source.  Thanks @quabug for the link.xml suggestion

**Comment by chung-ltkgame** [comment]
@svermeulen Hum. I try get attribute but Zenject Attribute is removed when build with il2cpp. I think the issue is in my library.

**Comment by svermeulen** [comment]
Fixed in version 7.0.0
